{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Functions\" data-toc-modified-id=\"Functions-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Functions</a></span><ul class=\"toc-item\"><li><span><a href=\"#1.-Write-a-function-that-returns-the-greater-of-two-numbers\" data-toc-modified-id=\"1.-Write-a-function-that-returns-the-greater-of-two-numbers-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>1. Write a function that returns the greater of two numbers</a></span></li><li><span><a href=\"#2.-Now-write-a-function-that-returns-the-largest-element-on-a-list\" data-toc-modified-id=\"2.-Now-write-a-function-that-returns-the-largest-element-on-a-list-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>2. Now write a function that returns the largest element on a list</a></span></li><li><span><a href=\"#3.-Write-a-function-that-sums-all-the-elements-on-a-list\" data-toc-modified-id=\"3.-Write-a-function-that-sums-all-the-elements-on-a-list-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>3. Write a function that sums all the elements on a list</a></span></li><li><span><a href=\"#4.-Write-another-function-that-multiplies-all-the-elements-on-a-list\" data-toc-modified-id=\"4.-Write-another-function-that-multiplies-all-the-elements-on-a-list-1.4\"><span class=\"toc-item-num\">1.4&nbsp;&nbsp;</span>4. Write another function that multiplies all the elements on a list</a></span></li><li><span><a href=\"#5.-Now-combine-those-two-ideas-and-write-a-function-that-receives-a-list-and-either-&quot;+&quot;-or-&quot;*&quot;-and-outputs-acordingly\" data-toc-modified-id=\"5.-Now-combine-those-two-ideas-and-write-a-function-that-receives-a-list-and-either-&quot;+&quot;-or-&quot;*&quot;-and-outputs-acordingly-1.5\"><span class=\"toc-item-num\">1.5&nbsp;&nbsp;</span>5. Now combine those two ideas and write a function that receives a list and either \"+\" or \"*\" and outputs acordingly</a></span></li><li><span><a href=\"#6.-Write-a-function-that-returns-the-factorial-of-a-number.\" data-toc-modified-id=\"6.-Write-a-function-that-returns-the-factorial-of-a-number.-1.6\"><span class=\"toc-item-num\">1.6&nbsp;&nbsp;</span>6. Write a function that returns the factorial of a number.</a></span></li><li><span><a href=\"#7.-Write-a-function-that-takes-a-list-and-returns-a-list-of-the-unique-values.\" data-toc-modified-id=\"7.-Write-a-function-that-takes-a-list-and-returns-a-list-of-the-unique-values.-1.7\"><span class=\"toc-item-num\">1.7&nbsp;&nbsp;</span>7. Write a function that takes a list and returns a list of the unique values.</a></span></li><li><span><a href=\"#8.-Write-a-function-that-returns-the-mode-of-a-list,-i.e.:-the-element-that-appears-the-most-times.\" data-toc-modified-id=\"8.-Write-a-function-that-returns-the-mode-of-a-list,-i.e.:-the-element-that-appears-the-most-times.-1.8\"><span class=\"toc-item-num\">1.8&nbsp;&nbsp;</span>8. Write a function that returns the mode of a list, i.e.: the element that appears the most times.</a></span></li><li><span><a href=\"#9.-Write-a-function-that-calculates-the-standard-deviation-of-a-list.\" data-toc-modified-id=\"9.-Write-a-function-that-calculates-the-standard-deviation-of-a-list.-1.9\"><span class=\"toc-item-num\">1.9&nbsp;&nbsp;</span>9. Write a function that calculates the standard deviation of a list.</a></span></li><li><span><a href=\"#10.-Write-a-function-to-check-if-a-string-is-a-pangram,-i.e.:-if-it-contains-all-the-letters-of-the-alphabet-at-least-once.-Mind-that-the-strings-may-contain-characters-that-are-not-letters.\" data-toc-modified-id=\"10.-Write-a-function-to-check-if-a-string-is-a-pangram,-i.e.:-if-it-contains-all-the-letters-of-the-alphabet-at-least-once.-Mind-that-the-strings-may-contain-characters-that-are-not-letters.-1.10\"><span class=\"toc-item-num\">1.10&nbsp;&nbsp;</span>10. Write a function to check if a string is a pangram, i.e.: if it contains all the letters of the alphabet at least once. Mind that the strings may contain characters that are not letters.</a></span></li><li><span><a href=\"#11.-Write-a-function-that-receives-a-string-of-comma-separated-words-and-returns-a-string-of-comma-separated-words-sorted-alphabetically.\" data-toc-modified-id=\"11.-Write-a-function-that-receives-a-string-of-comma-separated-words-and-returns-a-string-of-comma-separated-words-sorted-alphabetically.-1.11\"><span class=\"toc-item-num\">1.11&nbsp;&nbsp;</span>11. Write a function that receives a string of comma separated words and returns a string of comma separated words sorted alphabetically.</a></span></li><li><span><a href=\"#12.-Write-a-function-to-check-if-a-given-password-is-strong-(at-least-8-characters,-at-least-one-lower-case,-at-least-one-upper-case,-at-least-one-number-and-at-least-one-special-character).-It-should-output-True-if-strong-and-False-if-not.\" data-toc-modified-id=\"12.-Write-a-function-to-check-if-a-given-password-is-strong-(at-least-8-characters,-at-least-one-lower-case,-at-least-one-upper-case,-at-least-one-number-and-at-least-one-special-character).-It-should-output-True-if-strong-and-False-if-not.-1.12\"><span class=\"toc-item-num\">1.12&nbsp;&nbsp;</span>12. Write a function to check if a given password is strong (at least 8 characters, at least one lower case, at least one upper case, at least one number and at least one special character). It should output True if strong and False if not.</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On this lab we will put to practice some of the concepts we have learned on this past few days.\n",
    "\n",
    "`NOTE: On this lab you should try to write all the functions yourself using only the most basic of python syntax and without functions such as len, count, sum, max, min, in, etc. Give it a try. 🧑🏻‍💻👩🏻‍💻`\n",
    "\n",
    "The cell after each exercise contains a few tests to check if your function works as expected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run this cell\n",
    "\n",
    "from mod.testing import *\n",
    "import unittest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Write a function that returns the greater of two numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "def greater(a,b):\n",
    "    if (a > b):\n",
    "        print(a,\"is greater than\",b)\n",
    "    elif b > a:\n",
    "        print(b,\"is greater than\",a)\n",
    "\n",
    "pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "897 is greater than 45\n"
     ]
    }
   ],
   "source": [
    "greater(45, 897)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Now write a function that returns the largest element on a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def greatest(*args):\n",
    "    max = 0\n",
    "    for i in args:\n",
    "        if i > max:\n",
    "            max=i\n",
    "    return max\n",
    " \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8940"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greatest(45, 897, 5, 32, 8940, 22, 0, 89)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Write a function that sums all the elements on a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sum_all(*args):\n",
    "    a = 0\n",
    "    for i in args:\n",
    "        a = i + a\n",
    "\n",
    "    return a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "715"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_all(6, 64, 645)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Write another function that multiplies all the elements on a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mult_all(*args):\n",
    "    x = 1\n",
    "    for i in args:\n",
    "        x *= i\n",
    "   \n",
    "    return x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "168"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mult_all(4, 6, 7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Now combine those two ideas and write a function that receives a list and either \"+\" or \"*\" and outputs acordingly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def oper_all(arr, oper):\n",
    "    x = 0\n",
    "    if oper == (\"+\"):\n",
    "        x = sum_all(*arr)\n",
    "    else:\n",
    "        x = mult_all(*arr)\n",
    "    \n",
    "    return x\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "114"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oper_all([12, 23, 34, 45], \"+\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Write a function that returns the factorial of a number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def factorial(n):\n",
    "    p = 1\n",
    "    while n >= 1:\n",
    "        p = p * n\n",
    "        n = n - 1\n",
    "\n",
    "    return p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8841761993739701954543616000000"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "factorial(29)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Write a function that takes a list and returns a list of the unique values.\n",
    "\n",
    "`NOTE: You cannot use set. 🤔`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unique(arr):\n",
    "    list= []\n",
    "    for i in arr:\n",
    "        if i not in list:\n",
    "            list.append(i)\n",
    "    return list\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 3, 4]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique([1, 2, 3, 3, 4])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. Write a function that returns the mode of a list, i.e.: the element that appears the most times.\n",
    "`NOTE: You should not use count... 🧐`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mode_counter(arr):\n",
    "    y = {}\n",
    "    for a in arr:\n",
    "        if not a in y:\n",
    "         ###### INCOMPLETO ######\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 24 : Should be 24\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 3 : Should be 3\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 1 : Should be 1\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 4 : Should be 4\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 10 : Should be 10\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 12 : Should be 12\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 20 : Should be 20\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 22 : Should be 22\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 13 : Should be 13\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 23 : Should be 23\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 10 : Should be 10\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 11 : Should be 11\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 21 : Should be 21\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 11 : Should be 11\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 20 : Should be 20\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 17 : Should be 17\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 16 : Should be 16\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 9 : Should be 9\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 17 : Should be 17\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 18 : Should be 18\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 4 : Should be 4\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 23 : Should be 23\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 25 : Should be 25\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 8 : Should be 8\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 4 : Should be 4\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 22 : Should be 22\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 3 : Should be 3\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 5 : Should be 5\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 11 : Should be 11\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 20 : Should be 20\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 20 : Should be 20\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 16 : Should be 16\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 14 : Should be 14\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 17 : Should be 17\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 4 : Should be 4\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 20 : Should be 20\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 24 : Should be 24\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 20 : Should be 20\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 15 : Should be 15\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 16 : Should be 16\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 13 : Should be 13\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 14 : Should be 14\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 21 : Should be 21\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 18 : Should be 18\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 16 : Should be 16\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 14 : Should be 14\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 4 : Should be 4\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 12 : Should be 12\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 16 : Should be 16\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 8 : Should be 8\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 13 : Should be 13\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 18 : Should be 18\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 7 : Should be 7\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 8 : Should be 8\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 3 : Should be 3\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 15 : Should be 15\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 4 : Should be 4\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 25 : Should be 25\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 25 : Should be 25\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 1 : Should be 1\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 2 : Should be 2\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 21 : Should be 21\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 19 : Should be 19\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 6 : Should be 6\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 4 : Should be 4\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 20 : Should be 20\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 21 : Should be 21\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 25 : Should be 25\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 19 : Should be 19\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 3 : Should be 3\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 11 : Should be 11\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 5 : Should be 5\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 9 : Should be 9\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 11 : Should be 11\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 5 : Should be 5\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 6 : Should be 6\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 8 : Should be 8\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 13 : Should be 13\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 23 : Should be 23\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 5 : Should be 5\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 9 : Should be 9\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 2 : Should be 2\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 5 : Should be 5\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 2 : Should be 2\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 18 : Should be 18\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 21 : Should be 21\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 5 : Should be 5\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 11 : Should be 11\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 19 : Should be 19\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 17 : Should be 17\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 6 : Should be 6\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 15 : Should be 15\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 4 : Should be 4\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 20 : Should be 20\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 21 : Should be 21\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 18 : Should be 18\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 22 : Should be 22\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 3 : Should be 3\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 25 : Should be 25\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_mode.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\USUARIO\\Desktop\\1.3-lab-functions\\mod\\testing.py\", line 127, in runTest\n",
      "    self.assertEqual(fn(self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 20 : Should be 20\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.149s\n",
      "\n",
      "FAILED (failures=100)\n"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_mode(mode_counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9. Write a function that calculates the standard deviation of a list.\n",
    "`NOTE: Do not use any libraries or already built functions. 😉`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statistics import stdev\n",
    "def st_dev(arr):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will test your function \n",
    "test_stdev(st_dev)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 10. Write a function to check if a string is a pangram, i.e.: if it contains all the letters of the alphabet at least once. Mind that the strings may contain characters that are not letters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pangram(string):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will test your function \n",
    "test_pangram(pangram)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 11. Write a function that receives a string of comma separated words and returns a string of comma separated words sorted alphabetically.\n",
    "\n",
    "`NOTE: You may use sorted but not split and definitely no join! 🤪`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_alpha(string):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will test your function \n",
    "test_alpha(sort_alpha)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 12. Write a function to check if a given password is strong (at least 8 characters, at least one lower case, at least one upper case, at least one number and at least one special character). It should output True if strong and False if not.\n",
    "`Valid special characters: # @ ! $ % & ( ) ^ * [ ] { }`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_pass(string):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will test your function \n",
    "test_pass(check_pass)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  },
  "vscode": {
   "interpreter": {
    "hash": "5eba2f6daac758e0c4eb74cbbf3fd787eaf14bd08a71327f2f2275e3d00aa03b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
